### PR TITLE
Delete build_caffe2.sh, replace with build_libtorch.py

### PR DESCRIPTION
**Pull request by anderspapitto**
delete build_caffe2.sh, replace with build_libtorch.py as suggested by peter (and copy-pasted from his draft PR).  This ensures that all consumers of the torch CMake file go through as unified a path as possible.

In order to change the surrounding infrastructure as little as possible, I made some tweaks to enable build_pytorch_libs.sh to generate the test binaries relative to the current directory, rather than hardcoding to pytorch/build.

### CODE REVIEW

**Comment by anderspapitto** [comment]
@ezyang @goldsborough this is ready for your attention

**Review by goldsborough — APPROVED** [review]
LGTM

File: setup.py [hunk 1]
```diff
@@ -347,7 +347,14 @@ def build_libs(libs):
 
     my_env["BUILD_TORCH"] = "ON"
 
-    if subprocess.call(build_libs_cmd + libs, env=my_env) != 0:
+    try:
+        os.mkdir('build')
+    except OSError:
+        pass
+
+    kwargs = {'cwd': 'build'} if not IS_WINDOWS else {}
```
Comment (ezyang) [inline setup.py]: What's going on here?

**Review by ezyang — APPROVED** [review]
I like deleting files.

File: setup.py [hunk 2]
```diff
@@ -347,7 +347,14 @@ def build_libs(libs):
 
     my_env["BUILD_TORCH"] = "ON"
 
-    if subprocess.call(build_libs_cmd + libs, env=my_env) != 0:
+    try:
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply setup.py]: recommendations/best practices here are kind of mixed, i think. always doing the try-except version means you never have to worry that someone will be racing you

File: setup.py [hunk 3]
```diff
@@ -303,7 +303,7 @@ def build_libs(libs):
     if IS_WINDOWS:
         build_libs_cmd = ['tools\\build_pytorch_libs.bat']
     else:
-        build_libs_cmd = ['bash', 'tools/build_pytorch_libs.sh']
+        build_libs_cmd = ['bash', os.path.join('..', 'tools', 'build_pytorch_libs.sh')]
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply setup.py]: hmm. kind of ambivalent on this one. that adds another layer and build_libtorch.py is intended to do next to nothing, but on the other hand is another it is I guess more unified. I should probably follow up by making this change

File: .jenkins/pytorch/macos-test.sh [hunk 4]
```diff
@@ -56,8 +56,12 @@ test_cpp_api() {
   #
   CPP_BUILD="$PWD/../cpp-build"
   rm -rf $CPP_BUILD
-  mkdir -p $CPP_BUILD
-  WERROR=1 VERBOSE=1 tools/cpp_build/build_caffe2.sh "$CPP_BUILD"
+  mkdir -p $CPP_BUILD/caffe2
+
+  BUILD_LIBTORCH_PY=$PWD/tools/build_libtorch.py
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply .jenkins/pytorch/macos-test.sh]: no, just happens to be how I wrote it. I can follow up to change it around

File: setup.py (hunk 1 above)
Comment (anderspapitto) [reply setup.py]: for the linux/mac script, i've made the output go to the current directory - so now we have to make the current directory be `build` for things to go in the right place.

it's some localized messiness for the moment, but it should get cleaned up in passing as we merge build_pytorch_libs.{sh,bat} into the underlying cmake layer

**Comment by anderspapitto** [comment]
see also https://github.com/pytorch/pytorch/issues/10562 to track some related future improvements

File: setup.py (hunk 1 above)
Comment (ezyang) [reply setup.py]: It is always good when something is intended to be temporary to say so in a comment :)

File: .jenkins/pytorch/build.sh [hunk 5]
```diff
@@ -118,5 +118,9 @@ if [[ "$BUILD_TEST_LIBTORCH" == "1" ]]; then
   echo "Building libtorch"
   # NB: Install outside of source directory (at the same level as the root
   # pytorch folder) so that it doesn't get cleaned away prior to docker push.
-  WERROR=1 VERBOSE=1 tools/cpp_build/build_caffe2.sh "$PWD/../cpp-build"
+  BUILD_LIBTORCH_PY=$PWD/tools/build_libtorch.py
+  mkdir -p ../cpp-build/caffe2
+  pushd ../cpp-build/caffe2
```
Comment (goldsborough) [inline .jenkins/pytorch/build.sh]: Why the `caffe2/` suffix?

Comment (anderspapitto) [reply .jenkins/pytorch/build.sh]: that's just where everything was generated before this diff, and I don't have a reason to move them around

**Pull request closed without merging**